kubeadm operator (#1239)
kubernetes/enhancements: keps/sig-cluster-lifecycle/kubeadm/20190916-kubeadm-operator.md

consists of a complex sequence of tasks that should be manually executed and orchestrated across all the existing nodes
in the cluster.

Such a user experience is not ideal due to the error-prone nature of humans running commands. The manual approach
can be considered a blocker for implementing more complex workflows such as rotating certificate authorities,
modifying the settings of an existing cluster or any task that requires coordination of more than one Kubernetes node.

This KEP aims to address such problems by applying the operator pattern to kubeadm workflows. 

### Goals

- To allow declarative control of kubeadm operations that lead to "in place" mutations[1] of kubeadm generated
  artifacts. More specifically kubeadm artifacts are static pod manifests, certificates, kubeconfig file,
  bootstrap token, kubeadm generated configmap and secrets, while this proposal initially includes the following
  kubeadm workflows:
  - kubeadm upgrade
  - certificate renewal
  - certificate authority rotation (NEW)
  - change configuration in an existing cluster (NEW) 

> > [1] Please note that we are referring to "in place" mutations of kubeadm generated artifacts in order to highlight

Reviewer: CAPI recently merged tighter integration w/ cert-manager - I'm wondering if this footnote is outdated?

Author: At first sight, I don't see relations between cert-manager and the kubeadm-operator, but I prefer to double-check. Can you provide a link?

  - kubeadm upgrade
  - certificate renewal
  - change settings/configuration in an existing cluster (NEW) 
  - certificate authority rotation (NEW)
- To automate the orchestration of such mutations across the existing nodes in a cluster

> [1] Please note that we are referring to "in place" mutations of kubeadm generated artifacts in order to highlight the difference between the kubeadm operator and other SCL projects like [Cluster API](https://cluster-api.sigs.k8s.io/), which instead assume node/machines immutable

Considering the complexity of this topic, this document is expected to be subject to some iterations. The goal of the current iteration is to:

- Get initial approval on Summary and Motivation paragraphs
- To identify a semantic for defining “Actions” to be performed by the kubeadm-operator.
- To define how the kubeadm-operator should manage kubeadm workflows.
- To define how the users should interact with the kubeadm-operator, including also observability and error handling.
- To define how the kubeadm-operator should be deployed in a Kubernetes cluster.

> ### Non-Goals

Reviewer 2: I would explicitly mention that replacing `kubeadm` "raw" workflows is a non-goal, and if something, an enabler for the operator. If more actions are added, `kubeadm` will get those actions as well, and the operator will rely on them, but the user will always be able to run these commands in isolation in a manual fashion.

Author: Done!

- To provide or manage any infrastructure elements like e.g. machines, load balancer, etc.
- To manage/automate the kubeadm init/join workflows.
- To provide or manage any component/configuration file/artifacts not generated by kubeadm, with the only exception of the kubelet/kubeadm/kubectl binaries which are considering in scope as required by the upgrade workflow.

## Proposal

### User Stories

#### Story 1

> As a Kubernetes operator, I would like to be able to declaratively control upgrades in a systematic fashion.

Reviewer 2: Not a native speaker, but I think `systematic` here looks weird. Automatic, unattended? Same for the other occurences.

Reviewer 3: I think systematic makes sense if the intent is that there is some system to manage the configuration files that declare the workflows.

#### Story 2

As a Kubernetes operator, I would like to be able to declaratively control certificate renewal in a systematic fashion.

#### Story 3

As a Kubernetes operator, I would like to be able to declaratively control changes of the current cluster's settings in a systematic fashion.

#### Story 4

As a Kubernetes operator, I would like to be able to declaratively certificate authority rotation.

#### Story 5 
As a Kubernetes operator, I would like to control whether nodes are cordoned and drained for tasks or if they are performed "hot"

### Implementation Details/Notes/Constraints

The initial proposal for implementing the kubeadm operator can be summarized by the following sequence diagram; the goal of the current iteration is to validate/improve the proposed approach.

![](20190916-kubeadm-operator.png)

#### Action

> The first step of the sequence above is the user asking the kubeadm operator to perform an `Action`; `Action` is a new CRD that defines actions performed by the kubeadm operator in a high-level, end-user oriented format e.g. 

Author: Well this is really an interesting question.

When initially thinking about the operator, I was thinking of making the operator reconcile changes on the `ClusterConfiguration` object, assuming this as a description of the "desired state".

However, it quickly came to me that detecting changes to the `ClusterConfiguration` object and mapping such changes to the kubeadm workflows is far from trivial, and TBH, also a huge paradigm shift from what we are doing today. 
Are we, as a kubeadm maintainers, able to figure out all the knobs of this evolution?
Most importantly, are the kubeadm users ready for moving from today's "full control" to "total trust" to the kubeadm operator?

This is why I decided to reduce the scope and focus only on the part of the operator that orchestrates kubeadm workflows, the Actions.

And this is also the reason why I'm assuming that the operator should never automatically delete Action or Task, because outstanding observability of what is happening/what happened is a must for getting the users to trust the operator. 

Just to be clear, as soon as the orchestrates layer of the operator is in place and reliable, I'm totally OK in moving forward and start to design an higher-level controller that automatically generates actions according to the changes to the `ClusterConfiguration` object. 
I just feel we should get there step by step.

> The first step of the sequence above is the user asking the kubeadm operator to perform an `Action`; `Action` is a new CRD that defines actions performed by the kubeadm operator in a high-level, end-user oriented format e.g. 

Reviewer 4: [1] after giving this more thought, I think we will endup implementing bits of the Jobs + TTL controller if we want cleanup (which alpha IIRC).  I'd prefer us to refine the flows (short-lived, long-lived, short-lived with cleanup etc..) and see

cc @timothysc @neolit123

> The first step of the sequence above is the user asking the kubeadm operator to perform an `Action`; `Action` is a new CRD that defines actions performed by the kubeadm operator in a high-level, end-user oriented format e.g. 

Reviewer 5: Indeed, I can see all of your points here.
My personal preference is more on the declarative approach (hence actions are a bit uncanny to me). Diffing the kubeadm ClusterConfiguration in its entirety can be tricky and defining an action for each changed field is probably a doomed effort.
What I think, we should try to do is to start with an operator specific `ClusterConfiguration`. That is something on the lines of:
```yaml
apiVersion: operator.kubeadm.x-k8s.io/v1alpha1
kind: ClusterConfiguration
metadata:
     name: my-cluster
kubernetesVersion: 1.20.0
certificateRotationCount: 3
```

As you can see, this is not the kubeadm, but an operator specific `ClusterConfiguration`. It contains only fields, that we know how to tackle should they get changed. Some of the fields may not even be present in the kubeadm `ClusterConfiguration` (in this example, `certificateRotationCount`).
That way we still end with a declarative approach and a narrow set of actions for a start.

In addition to this, I do have the same concern as @yastij, that we may end up re-implementing large chunks of the Job and/or DaemonSet controllers.

> The first step of the sequence above is the user asking the kubeadm operator to perform an `Action`; `Action` is a new CRD that defines actions performed by the kubeadm operator in a high-level, end-user oriented format e.g. 

Reviewer 3: Ah so a wrapper on top of a cluster configuration? My expectation here is that when this operator is installed it will generate the object that reflects the state of current configuration. Is that right? Then, a modification to this object will result in some actions on the cluster in an attempt to get the cluster to match the supplied configuration.

I think that's a reasonable way to build up to all the features we'd need to support and I would be in favor of this approach.

> The first step of the sequence above is the user asking the kubeadm operator to perform an `Action`; `Action` is a new CRD that defines actions performed by the kubeadm operator in a high-level, end-user oriented format e.g. 

Reviewer 5: > Ah so a wrapper on top of a cluster configuration?

Depends on what you call "wrapper on top of a cluster configuration". The full `kubeadm.ClusterConfiguration` don't need to be exposed, but only a subset of fields from it. The new `operator.ClusterConfiguration` should also have some fields to deal with concepts, that are not reflected in the `kubeadm.ClusterConfiguration`. Those concept are usually backed by kubeadm actions. Certificate rotation is one such thing.

For example, the above proposition is roughly going to match the following struct:

```golang
type ClusterConfiguration struct {
	TypeMeta
	
	// Here, ObjectMeta.Name is going to correspond to kubeadmapiv1beta2.ClusterConfiguration.ClusterName
	ObjectMeta
	
	// 1:1 correspondence to kubeadmapiv1beta2.ClusterConfiguration.KubernetesVersion
	// A change here is going to trigger upgrade on the machines to the target version
	KubernetesVersion string
	
	// A new field, not present in kubeadmapiv1beta2.ClusterConfiguration.
	// This is a number, that can only be bumped up. A change to this will trigger certificate rotation.
	CertificateRotationCound int
}
```

> My expectation here is that when this operator is installed it will generate the object that reflects the state of current configuration. Is that right? Then, a modification to this object will result in some actions on the cluster in an attempt to get the cluster to match the supplied configuration.

That's my vision of where things should be going too.

> The first step of the sequence above is the user asking the kubeadm operator to perform an `Action`; `Action` is a new CRD that defines actions performed by the kubeadm operator in a high-level, end-user oriented format e.g. 

Reviewer 3: If we're tossing out ideas, I know this has no relevance to the PR anymore, but it would be cool if the object had a field like `CertificateExpiration` and showed you when the certificates were going to expire. Then you could set the field to a future date and your certs would rotate to get a new expiration date that matched your desired expiration date, perhaps with an "autoRenew: bool" field on that always never let your certificates expire.

I recognize these behaviors can all be expressed with `Action`s, however.

> The first step of the sequence above is the user asking the kubeadm operator to perform an `Action`; `Action` is a new CRD that defines actions performed by the kubeadm operator in a high-level, end-user oriented format e.g. 

Author: @chuckha @rosti @yastij 
Point taken.
Please consider that we are approaching this incrementally, and mi suggestion is to initially focus on the orchestration layer. In the meantime, we can discuss how to best approach the higher level reconciliation loop, that will be implemented in next iterations.

  name: action-sample
spec:
  upgrade: 
      targetVersion: v1.16.0
```

or 

```yaml
apiVersion: operator.kubeadm.x-k8s.io/v1alpha1
kind: Action
metadata:
  name: action-sample
spec:
  renewCertificates: {}
```
#### Action Controller and TaskGroup
The `Action controller` is a component of the kubeadm operator that manages the  `Action` CRD and implements all the domain-specific knowledge about the kubeadm workflows e.g. it knows that the certificate-renewal workflow requires to run `kubeadm alpha certs renew` on all the control plane nodes in the cluster.

As a consequence, the  `Action controller`  generates one or more `TaskGroup` objects, each describing a `TaskTemplate` object - e.g. run `kubeadm alpha certs renew` -  targeting a set of nodes matching a given `NodeSelector` - e.g. all nodes with `node-role.kubernetes.io/master` label -.

```yaml
apiVersion: operator.kubeadm.x-k8s.io/v1alpha1
kind: TaskGroup
metadata:
  name: taskgroup-sample2
  ownerReference:
      ...
spec:
  nodeSelector: 
    matchLabels:
      node-role.kubernetes.io/master: ""
  selector:
    matchLabels:
      app: a
  template:
    metadata:
      labels:
        app: a
    alphaCertsRenew: {}
```

> It is important to notice that the  `Action controller` is going to always generate `TaskGroup` in sequential order - only after the current `TaskGroup` is completed successfully a new one will be generated.

Reviewer 2: What happens if the current `TaskGroup` does not succeed? Assuming an upgrade, if I understood correctly:

1. TaskGroup created for control plane nodes (`node-role.kubernetes.io/master: ""`)
2. When 1. has succeeded, TaskGroup created for worker nodes (unclear how they'll be targeted if not explicitly labeled -- with a `matchExpression` maybe?)

If 1. fails, will it be retried?; do you foresee a retrial count limit?. Will the `TaskGroup` be eventually removed?

Author: > What happens if the current TaskGroup does not succeed?

The current working assumption is to block the execution immediately after a task fail, and to not take initiative automatically. I'm working on a possible UX to let the user to choose if to retry or to skip the failed task.

> unclear how they'll be targeted if not explicitly labeled -- with a matchExpression maybe

matchExpression is an option, even if not currently supported by controller-tools. 
I donìt have a veto also on requiring additional labels for the operator (might be only during alpha)

Reviewer 5: Still, we can try to do away with the Task objects (possibly even with Task Groups) and leverage DaemonSets and Jobs where part of these problems have been solved and kubectl is your friend.

Author: I have described pros and cons about using bare DaemonSets and Jobs the alternatives paragraphs, and TBH I don't that going on this path is a good idea, because we will be stuck as soon as we have a small requirement that deviates from DaemonSets and Jobs abstractions. And as far as I am concerned, we are already in this case for operability requirements (e.g. dry-run, pause, error management)

Reviewer 4: [1]

#### Observability and Operability
The kubeadm operator is designed to take responsibility for actions that as of today are in charge of users, and we want to ensure that the users can feel comfortable in handling over such responsibility.

This goal influences the kubeadm operator design as described in the following paragraphs

##### Execution order

The kubeadm operator is going to execute `Action`'s `Task`s in a predictable and consistent order:

- `TaskGroup` will be generated in the order encoded in each action and always executed in sequential order.
- If a `TaskGroup` uses `sequential` policy, `Task` will be executed according to the alphabetical order of Nodes.
- Additional `TaskGroup` policies introduced in future, are expected to introduce similar guarantees. 

Additionally, it will be possible to execute actions in `dry-run` mode, thus allowing users to have a preview of the resulting task order.

##### Execution modes

The kubeadm operator is going to execute`Task`s acconding to the order described above, and by default a new `Task` will be automatically started as soon as the previous task completes.

> At any time during the execution, the user will be allowed to pause the `Action` execution using the new `kubeadm operator pause` command.

Reviewer 2: ```suggestion
At any time during the execution, the user will be allowed to pause the next `Action` execution using the new `kubeadm operator pause` command.
```

Author: Actually the idea is to not pause the next action, but the next task within the same action.
I rephrased to make this more clear

Similarly, it will be possible to restart the execution with `kubeadm operator restart`.

It will be also possible to execute action in `Controlled` mode; with this option, the kubeadm operator to automatically pause before actually executing each task; this will allow the user to check the task details and decide whether to proceed or not.

> NB. Once a `Task` is handled over to kubeadm, it is not possible to pause its execution; that means that `kubeadm operator pause` does not affect running `Task`, but it prevents new new `Task` to be started. 

##### Error management

In case of errors, the kubeadm operator will immediately pause `Action` execution.

During the proof of concept we will explore options for allowing retry of failed tasks and/or restart and skipping failed task.

#### Extensibility

The kubeadm operator provides high level `Action` encoding the knowledge of how kubeadm workflows should be performed in the form of `TaskGroup` and `Task`. 

However, such out of the box experience might not be fit for advanced use cases, custom clusters etc.

The kubeadm operator will provide a toolbox for allowing users to address such use cases as well:

> - It will be introduced a `CustomAction` type, allowing the user to define custom lists of `TaskGroup` and `Task`.

Reviewer 2: I think this somewhat fails out of the scope of this RFC? Once that the kubeadm operator is in place we can check for extensibility and how to customize it in the best way possible, once that we have feedback from users. I would try to aim for a MVP regarding the operator, then iterate.

Author: Added a sentence to formalize this is a stretched objective for the MVP.
However, I'm confident we can get this almost for free and it helps for testing the operator itself, so currently I'm still for getting this at the very beginning

- It will be possible to create manually "detached" `TaskGroup` and/or `Task` (without a parent `Action`), thus allowing to the user further extension points.

> ### Risks and Mitigations

Reviewer 2: If a container in a pod restarts the kubelet (e.g. as part of an upgrade process), what I have seen in the past is that we won't have feedback of the success/failure of that container execution, because the container itself was restarting the kubelet. I think this is a risk, that certain operations because of their nature are not properly detected, and because of them running containerized by the kubelet itself might not be trivial to introspect.

Author: Thanks for the head up.
I'm not 100% sure this is our case because we are going to use a long-running container as an agent for running commands on demand, but this is something to clarify asap as we start the implementation

Reviewer 2: Yes, it also depends on the implementation: in my case I was looking for the success/failure of a job that was restarting the kubelet. In this case, if we annotate the custom resource with success right before restarting the kubelet we should be safe (the only thing we cannot determine would be the restart of the kubelet itself).

We can find out during the implementation though.

[1] A hierarchy of CRD objects - `Action`,  `TaskGroup` , `Task` - provides a clean semantics and helps in having a good separation of concerns between  controllers, but as back side, it introduces new problems like orphan management.

Those conditions can arise only if the users accidentally or intentionally creates or deletes `TaskGroup`/`Task` while executing one `Action`.

Such behaviour should be avoided or executed only by very advanced users, but there is no way to ensure this can't happen.

During the proof of concept we will explore options for detecting such conditions and immediately pause execution in case of inconsistencies.

Additionally, we are going to ensure that completed `Action` will not reconcile anymore in order to prevent accidental execution of tasks.

[2] The `Task controller` require to be executed in privileged mode, and this might raise concern in terms of security.

> However, considering that the kubeadm operator does not require more privileges than the admin executing the same command manually, we consider that we are not introducing new security risks with this proposal.

Reviewer 2: Since the operator will usually perform "one shot" operations at certain points in time, I think it would be interesting that `kubeadm` allows you to:

1. Deploy the operator (and daemonset), use it to perform cluster-wide kubeadm workflows
2. Uninstall the operator (and the daemonset with it)

In my opinion this approach has no security drawbacks, because the operator is not running the whole time on the cluster, only on demand. While I see the advantages WRT for example, certificate renewal, I think we can at least explore this option in the future (probably out of scope for this KEP).

Author: I added this idea as an option for reducing the risk surface, but assuming the Task Controller is already there on nodes helps in simplifying the first MVP

Reviewer 5: The `Task controller` resembles something in the middle of a regular Kubernetes Job controller and DaemonSet controller. We may also consider reusing some of those (so we don't reinvent k8s itself).

---
title: Kubeadm operator
authors:
  - "@fabriziopandini"
owning-sig: sig-cluster-lifecycle
participating-sigs:
  - sig-cluster-lifecycle=
reviewers:
  - "@neolit123"
  - "@rosti"
  - "@ereslibre"
  - "@detiber"
  - "@vincepri"
  - "@yastij"
  - "@chuckha"
approvers:
  - "@timothysc"

>   - "@luxas"

Reviewer 3: is luxas still around for this? I can't remember his timeline

Author: He show up sometimes at the sig meeting, and I would prefer to keep it into the loop also for kubeadm related KEPs.

- [Design Details](#design-details)
  - [Test Plan](#test-plan)
  - [Graduation Criteria](#graduation-criteria)
  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
  - [Version Skew Strategy](#version-skew-strategy)
- [Implementation History](#implementation-history)
- [Drawbacks](#drawbacks)
- [Alternatives](#alternatives)
- [Infrastructure Needed](#infrastructure-needed)
<!-- /toc -->

## Release Signoff Checklist

- [x] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
- [x] KEP approvers have set the KEP status to `implementable`
- [x] Design details are appropriately documented
- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
- [ ] Graduation criteria is in place
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

## Summary

> Kubeadm operator would like to enable declarative control of kubeadm workflows, automating the execution and the orchestration of such tasks across existing nodes in a cluster.

Reviewer 3: can the list of tasks be specified or some examples be given in the summary?

Author: The KEP specification recommend short Summary; instead in the motivation I'm giving more detail and linking the kubeadm upgrade page as a example

## Motivation

Kubeadm binary can execute actions only on the machine where it is running e.g. it is not possible to execute actions on other nodes, to copy files across nodes, etc.

> As a consequence, most of the kubeadm workflows, like e.g. [kubeadm upgrade](https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-upgrade/), consists of a complex sequence of tasks that should be manually executed/orchestrated across all the existing nodes in the cluster.

Reviewer 3: ```suggestion
As a consequence, most of the kubeadm workflows, like [`kubeadm upgrade`](https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-upgrade/), consists of a complex sequence of tasks that should be manually executed and orchestrated across all the existing nodes in the cluster.
```

> Such a user experience is not ideal, potentially error-prone, and this approach can be considered also a limitation/blocker for implementing more complex workflows like e.g. rotating certificate authorities, change the settings of an existing cluster etc.

Reviewer 3: ```suggestion
Such a user experience is not ideal due to the error-prone nature of humans running commands. The manual approach can be considered a blocker for implementing more complex workflows such as rotating certificate authorities, modifying the settings of an existing cluster or any task that requires coordination of more than one Kubernetes node.
```

This KEP aims to address such problems by applying the operator pattern to kubeadm workflows. 

### Goals

> - To allow declarative control of kubeadm action that leads to "in place" mutations[1] of kubeadm generated artifacts. More specifically this initially targets includes kubeadm workflows:

Reviewer 3: ```suggestion
To allow declarative control of kubeadm actions that lead to "in place" mutations[1] of kubeadm generated artifacts. More specifically this initially includes the following kubeadm workflows:
```

Is it worth documenting kubeadm artifacts? (manifests, cas, admin kubeconfig, bootstrap token, kubeadm configmap)

Author: Good suggestion, done!

>   - kubeadm upgrade

Reviewer 3: ```suggestion
  - cluster component upgrades
```

Author: I prefer kubeadm upgrade to be consistent with the kubeadm documentation in kubernetes.io and the rest of this document

  - certificate renewal

>   - change settings/configuration in an existing cluster (NEW) 

Reviewer 3: ```suggestion
  - change configuration in an existing cluster (NEW) 
```

>   - certificate authority rotation (NEW)

Reviewer 3: Could move this up to be next to certificate renewal in an attempt to group workflows?

Author: done!

> - To automate the orchestration of such mutations across the existing nodes in a cluster

Reviewer 3: This feel like a rehash of the original goal, not an independent kubeadm workflow

Author: removed!

> > [1] Please note that we are referring to "in place" mutations of kubeadm generated artifacts in order to highlight the difference between the kubeadm operator and other SCL projects like [Cluster API](https://cluster-api.sigs.k8s.io/), which instead assume node/machines immutable

Reviewer 3: ```suggestion
> [1] Please note that we are referring to "in place" mutations of kubeadm generated artifacts in order to highlight the difference between the kubeadm operator and other SCL projects like [Cluster API](https://cluster-api.sigs.k8s.io/), which instead assume nodes and underlying machines are immutable.
```

Considering the complexity of this topic, this document is expected to be subject to some iterations. The goal of the current iteration is to:

- Get initial approval on Summary and Motivation paragraphs
- To identify a semantic for defining “Actions” to be performed by the kubeadm-operator.
- To define how the kubeadm-operator should manage kubeadm workflows.
- To define how the users should interact with the kubeadm-operator, including also observability and error handling.
- To define how the kubeadm-operator should be deployed in a Kubernetes cluster.

### Non-Goals

> - To provide or manage any infrastructure elements like e.g. machines, load balancer, etc.

Reviewer 3: ```suggestion
- To provide or manage any infrastructure elements such as underlying machines, load balancers, storage, etc.
```

> - To manage/automate the kubeadm init/join workflows.

Reviewer 3: ```suggestion
- To manage and automate the kubeadm init and join workflows.
```

> - To provide or manage any component/configuration file/artifacts not generated by kubeadm, with the only exception of the kubelet/kubeadm/kubectl binaries which are considering in scope as required by the upgrade workflow.

Reviewer 3: ```suggestion
- To provide or manage any component, configuration file, or artifact *not* generated by kubeadm. The only exception are the kubelet, kubeadm, and kubectl binaries which are considered in-scope as required by the upgrade workflow.
```

> during implementation of the first proof of concept we might consider to extend `Task` template semantic in order to allow to execute more than one command.

> during implementation of the first proof of concept we might consider to implement a lock mechanisms for prevventing concurrent execution of more than one `Action` at any time.

> Alternative name for the `TaskGroup` might be `TaskBatch` (batch = a quantity or consignment of goods produced at one time).
> Instead usage of `TaskSet` or `TaskDeployment` was discarded because of its generally imply something that creates Pods, which it is not the case.

#### TaskGroupController and Task
The `TaskGroup controller` will be responsible for managing the  `TaskGroup` CRD and implements all the logic for generating `Task` from a `TaskGroup` according to the current topology of the Kubernetes cluster.

Each generated `Task` is expected to target a specific `Node` in the cluster.

```yaml
apiVersion: operator.kubeadm.x-k8s.io/v1alpha1
kind: Task
metadata:
  name: task-sample-control-plane-1
  labels:
    app: a
spec:
  nodeName: control-plane-1

>   alphaCertsRenew: {}

Reviewer 4: Does `alphaCertsRenew` include any fields ?

Maybe we should add the types in the design

Author: I don't have a design ready for all the Task, but I assume they will be similar to the command CLI.

However, at this stage, I prefer to get a high-level agreement on how the operator is goin to orchestrate workflows (with `Action`, `TaskGroup` and `Task`) before getting into the detailed API review

Reviewer 4: sgtm

kind: Task
metadata:
  name: task-sample-control-plane-2
  labels:
    app: a
spec:
  nodeName: control-plane-2
  alphaCertsRenew: {}
```

The `TaskGroup controller` is going to generate `Task` according to a policy that by default is sequential order - only after the current `Task` is completed successful a new one will be generated.

In future releases of the kubeadm operator also additional policy could be added for supporting e.g. parallel execution of `Task` or disruption budget (max N `Task` in executed at the same time).

In any case, the `TaskGroup controller` is going to ensure a predictable execution order based on the alphabetical order of `Node`s in the scope of a `TaskGroup`.

#### TaskController

The `Task controller` is the last component of the kubeadm operator. 

The main characteristic of this controller is the fact that this controller is deployed on all the nodes in the cluster using a `DaemonSet`, and that each instance is responsible for managing `Task` targeting a specific node.

`Task` reconciliation triggers the execution of kubeadm commands on such node; in order to do so, `Task controller` require to be executed in privileged mode.

> #### Observability and Operability

Reviewer 4: I think we should add a dedicated subsection for observability, for log management event. It shouldn't be detailed, but just discuss how would a user troubleshoot a failed action/taskGroup

Author: Good suggestion, done!

The kubeadm operator is designed to take responsibility for actions that as of today are in charge of users, and we want to ensure that the users can feel comfortable in handling over such responsibility.

This goal influences the kubeadm operator design as described in the following paragraphs

##### Execution order

The kubeadm operator is going to execute `Action`'s `Task`s in a predictable and consistent order:

> - `TaskGroup` will be generated in the order encoded in each action and always executed in sequential order.

Reviewer 4: Maybe I missed this, but how will the user specify a custom order ?

Author: The user can't specify order because each `Action` is going to generate a "predefined", ordered list of `TaskGroup`.

This paragraph is stressing this point because this ensures predictable behavior.

Reviewer 4: we can discuss this later when/if we will support custom workflows

This KEP aims to address such problems by applying the operator pattern to kubeadm workflows. 

### Goals

- To allow declarative control of kubeadm actions that lead to "in place" mutations[1] of kubeadm generated artifacts. More specifically kubeadm artifacts are static pod manifests, certificates, kubeconfig file, bootstrap token, kubeadm generated configmap and secrets, while this proposal initially includes the following kubeadm workflows:
  - kubeadm upgrade
  - certificate renewal
  - certificate authority rotation (NEW)
  - change configuration in an existing cluster (NEW) 

> [1] Please note that we are referring to "in place" mutations of kubeadm generated artifacts in order to highlight the difference between the kubeadm operator and other SCL projects like [Cluster API](https://cluster-api.sigs.k8s.io/), which instead assume nodes and underlying machines are immutable.

Considering the complexity of this topic, this document is expected to be subject to some iterations. The goal of the current iteration is to:

- Get initial approval on Summary and Motivation paragraphs

> - To identify a semantic for defining “Actions” to be performed by the kubeadm-operator.

Author: done

#### Story 2

As a Kubernetes operator, I would like to be able to declaratively control certificate renewal in a systematic fashion.

#### Story 3

As a Kubernetes operator, I would like to be able to declaratively control changes of the current cluster's settings in a systematic fashion.

#### Story 4

As a Kubernetes operator, I would like to be able to declaratively rotate my certificate authorities.

#### Story 5 
As a Kubernetes operator, I would like to control whether nodes are cordoned and drained for tasks or if they are performed without disruption to the workloads on the node, referred to as a "hot" update.

### Implementation Details/Notes/Constraints

The initial proposal for implementing the kubeadm operator can be summarized by the following sequence diagram; the goal of the current iteration is to validate/improve the proposed approach.

![](20190916-kubeadm-operator.png)

#### Action
The first step of the sequence above is the user asking the kubeadm operator to perform an `Action`; `Action` is a new CRD that defines actions performed by the kubeadm operator in a high-level, end-user oriented format e.g. 

```yaml
apiVersion: operator.kubeadm.x-k8s.io/v1alpha1

> kind: Action

Author: point taken
It is a declared objective of the first iteration to discover a good semantic for expressing operations. Any suggestion is more than welcome!

#### Story 4

As a Kubernetes operator, I would like to be able to declaratively rotate my certificate authorities.

#### Story 5 

> As a Kubernetes operator, As a Kubernetes operator, I would like to control whether nodes are cordoned and drained for tasks or if they are performed without disruption to the workloads on the node, referred to as a "hot" update.

Reviewer 4: `As a Kubernetes operattor, I would...`
